### PR TITLE
planning: open space implement slack osqp

### PR DESCRIPTION
**Pull request by deidaraho**

### CODE REVIEW

File: modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+/*
+ * @file
+ */
+#include <algorithm>
+#include "modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.h"
```
Comment (xiaoxq) [inline modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc]: Put it at the top.

File: modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+      int tmp_c = static_cast<int>(i);
+      int tmp_r = static_cast<int>(P_indices[idx]);
+      constraint_A_(tmp_r, tmp_c) = static_cast<float>(P_data[idx]);
+    }
+  }
+}
+
+bool DualVariableWarmStartSlackOSQPInterface::optimize() {
+  int kNumParam = num_of_variables_;
+  int kNumConst = num_of_constraints_;
```
Comment (xiaoxq) [inline modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc]: Only const variables follow the `kXxx` name pattern.

File: modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+
+  for (int i = 0; i < horizon_ + 1; ++i) {
+    for (int j = 0; j < obstacles_num_; ++j) {
+      max_s = std::max(max_s, slacks_(j, i));
+      min_s = std::min(min_s, slacks_(j, i));
+    }
+  }
+
+  LOG(INFO) << "max_s: " << std::to_string(max_s);
+  LOG(INFO) << "min_s: " << std::to_string(min_s);
```
[the remark this answers is not in the document]
Comment (Capri2014) [reply modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc]: Also set this as DEBUG information instead

File: modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+    }
+  }
+
+  // Problem settings
+  OSQPSettings* settings =
+      reinterpret_cast<OSQPSettings*>(c_malloc(sizeof(OSQPSettings)));
+
+  // Define Solver settings as default
+  osqp_set_default_settings(settings);
+  settings->alpha = 1.0;  // Change alpha parameter
```
Comment (Capri2014) [inline modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc]: load these from setting

File: modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_problem.cc [hunk 5]
```diff
@@ -51,6 +51,28 @@ bool DualVariableWarmStartProblem::Solve(
...
+
+      auto t_end = cyber::Time::Now().ToSecond();
+      ADEBUG << "Dual variable warm start solving time in second : "
+             << t_end - t_start;
+
+      solver_flag = true;
+    } else {
+      AWARN << "dual warm up fail.";
+      ptop->get_optimization_results(l_warm_up, n_warm_up);
+      solver_flag = false;
```
[the remark this answers is not in the document]
Comment (deidaraho) [reply modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_problem.cc]: > Simply let `solver_flag = ptop->optimize();` and remove `succ`.

fixed

[the remark this answers is not in the document]
Comment (deidaraho) [reply modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_problem.cc]: > Simply let `solver_flag = ptop->optimize();` and remove `succ`.

Fixed

File: modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_problem.cc [hunk 6]
```diff
@@ -51,6 +51,28 @@ bool DualVariableWarmStartProblem::Solve(
...
+    }
+  } else if (planner_open_space_config_.dual_variable_warm_start_config()
+          .qp_format() == SLACKQP) {
+    DualVariableWarmStartSlackOSQPInterface* ptop =
+        new DualVariableWarmStartSlackOSQPInterface(
```
[the remark this answers is not in the document]
Comment (deidaraho) [reply modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_problem.cc]: > I don't know the context. But this seems a memory leak.

Fixed

File: modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.h [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+  void assemble_constraint(std::vector<c_float>* A_data,
+                           std::vector<c_int>* A_indices,
+                           std::vector<c_int>* A_indptr);
+
+  void assembleA(const int r, const int c, const std::vector<c_float>& P_data,
+                 const std::vector<c_int>& P_indices,
+                 const std::vector<c_int>& P_indptr);
+
+  void check_solution(const Eigen::MatrixXd& l_warm_up,
```
[the remark this answers is not in the document]
Comment (deidaraho) [reply modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.h]: > Use camel-case.

Fixed

File: modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (deidaraho) [reply modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc]: > Use AINFO.



> Also set this as DEBUG information instead

Both fixed

File: modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc (hunk 1 above)
Comment (deidaraho) [reply modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc]: > Put it at the top.

Fixed

File: modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc (hunk 2 above)
Comment (deidaraho) [reply modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc]: > Only const variables follow the `kXxx` name pattern.

Fixed

File: modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc (hunk 4 above)
Comment (deidaraho) [reply modules/planning/open_space/trajectory_smoother/dual_variable_warm_start_slack_osqp_interface.cc]: > load these from setting

I wrapped osqp configs into open space protobuf.

**Pull request merged**